### PR TITLE
PowerShell module publishing improvements

### PR DESCRIPTION
**Pull request by splatteredbits**

### CODE REVIEW

**Comment by splatteredbits** [comment]
@webmd-health-services/devops Build is finally green. Ready for review.

File: Whiskey/en-US/about_Whiskey_PublishPowerShellModule_Task.help.txt [hunk 1]
```diff
@@ -9,20 +9,28 @@ SUMMARY
...
 
-    If a repository named `RepositoryName`, doesn't exist, one is registered and its source and publish locations are both set to `RepositoryUri` (using the `PowerShellGet` modules's `Register-PSRepository` function.
+    If a repository named `RepositoryName` doesn't exist, you must supply the `RepositoryUri` parameter so that the task can register the repository. The `RepositoryUri` parameter is passed to the `Register-PSRepository` function's `PublishLocation` and `SourceLocation` parameters.
 
-    Before publishing, the `ModuleVersion` property in the module manifest is updated to be the version of the current build, minus any prerelease or build metadata.
+    Before publishing, the `ModuleVersion` property in the module manifest is updated to be the version of the current build, minus any prerelease or build metadata and the `Prerelease` property in the `PrivateData.PSData` section is updated to be the current prerelease metadata. Make sure you use prerelease metadata in the correct semver format for the repository to which you're publishing. The task uses regular expressions to find and replace the module version and prerelease. It looks for (excluding the comments):
```
Comment (DecoyJoe) [inline Whiskey/en-US/about_Whiskey_PublishPowerShellModule_Task.help.txt]: Should we offer a way to disable this functionality? Sometimes people write their own version automation or use tools that do automated version management during builds, which this functionality would interfere with.

File: Test/Add-WhiskeyApiKey.Tests.ps1 [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        [string]
+        $ID,
+
+        [Parameter(Mandatory)]
+        [string]
+        $WithValue,
+
+        [Parameter(Mandatory)]
+        [Switch]
+        $Exists
```
[the remark this answers is not in the document]
Comment (splatteredbits) [reply Test/Add-WhiskeyApiKey.Tests.ps1]: I use it to make the assertion read better, e.g, `ThenApiKey 'ID' -WithValue 'Fubar' -Exists`.

File: Whiskey/Tasks/Publish-WhiskeyPowerShellModule.ps1 [hunk 3]
```diff
@@ -67,27 +60,56 @@ function Publish-WhiskeyPowerShellModule
...
         return
     }
 
     $manifest = Get-Content $manifestPath
-    $versionString = "ModuleVersion = '{0}.{1}.{2}'" -f ( $TaskContext.Version.SemVer2.Major, $TaskContext.Version.SemVer2.Minor, $TaskContext.Version.SemVer2.Patch )
-    $manifest = $manifest -replace "ModuleVersion\s*=\s*('|"")[^'""]*('|"")", $versionString
+    $versionString = 'ModuleVersion = ''{0}.{1}.{2}''' -f ( $TaskContext.Version.SemVer2.Major, $TaskContext.Version.SemVer2.Minor, $TaskContext.Version.SemVer2.Patch )
+    $manifest = $manifest -replace 'ModuleVersion\s*=\s*(''|")[^''"]*(''|")', $versionString
+    $prereleaseString = 'Prerelease = ''{0}''' -f $TaskContext.Version.SemVer2.Prerelease  
+    $manifest = $manifest -replace 'Prerelease\s*=\s*(''|")[^''"]*(''|")', $prereleaseString
```
[the remark this answers is not in the document]
Comment (splatteredbits) [reply Whiskey/Tasks/Publish-WhiskeyPowerShellModule.ps1]: No. I did that on purpose. If someone wants pre-release, they'll need to make sure it is uncommented. Likewise, if they don't want it, they should be able to comment it out.

File: Whiskey/Tasks/Publish-WhiskeyPowerShellModule.ps1 [hunk 4]
```diff
@@ -67,27 +60,56 @@ function Publish-WhiskeyPowerShellModule
...
+    }
+    if( (Test-Path -Path 'variable:InformationPreference') )
+    {
+        $commonParams['InformationAction'] = $InformationPreference
+    }
+
+    Get-PackageProvider -Name 'NuGet' -ForceBootstrap @commonParams | Out-Null
 
-    if( -not (Get-PSRepository -Name $repositoryName -ErrorAction Ignore) )
+    if( -not (Get-PSRepository -Name $repositoryName -ErrorAction Ignore @commonParams) )
```
[the remark this answers is not in the document]
Comment (splatteredbits) [reply Whiskey/Tasks/Publish-WhiskeyPowerShellModule.ps1]: You can publish to an untrusted repository. You just can't read from an untrusted repository without confirmation.

File: Whiskey/en-US/about_Whiskey_PublishPowerShellModule_Task.help.txt (hunk 1 above)
Comment (splatteredbits) [reply Whiskey/en-US/about_Whiskey_PublishPowerShellModule_Task.help.txt]: Good idea. I entered issue https://github.com/webmd-health-services/Whiskey/issues/239 to get this functionality added.

File: Whiskey/en-US/about_Whiskey_PublishPowerShellModule_Task.help.txt [hunk 5]
```diff
@@ -9,20 +9,28 @@ SUMMARY
...
     
-    You must supply the the path to the module's directory with the `Path` property, the URI of the repository with the `RepositoryUri` property, the name of the repository with the `RepositoryName` property, and the ID of the API key to use with the `ApiKeyID`.
+    You must supply the the path to the module's directory with the `Path` property, the name of the repository with the `RepositoryName` property, and the ID of the API key to use with the `ApiKeyID`.
 
-    If a repository named `RepositoryName`, doesn't exist, one is registered and its source and publish locations are both set to `RepositoryUri` (using the `PowerShellGet` modules's `Register-PSRepository` function.
+    If a repository named `RepositoryName` doesn't exist, you must supply the `RepositoryUri` parameter so that the task can register the repository. The `RepositoryUri` parameter is passed to the `Register-PSRepository` function's `PublishLocation` and `SourceLocation` parameters.
```
[the remark this answers is not in the document]
Comment (splatteredbits) [reply Whiskey/en-US/about_Whiskey_PublishPowerShellModule_Task.help.txt]: Changed to read "If a repositry named \`RepositoryName\` isn't registered".

**Pull request merged**